src/Database/PG/Query/Class.hs

```diff
@@ -448,69 +451,106 @@ class ToPrepArgs a where
 instance ToPrepArgs () where
   toPrepArgs _ = []
 
+-- | Convert a value into a prepared query argument.
+toPrepArg :: forall a. ToPrepArg a => a -> PrepArg
+toPrepArg x = (oid @a, fmap go (encoder x))
+  where
+    go e = (PE.encodingBytes e, PQ.Binary)
+
+-- | Tools for preparing values to be query parameters. A prepared query
+-- parameter (or 'PrepArg') contains an Oid (allowing Postgres to deduce its
+-- type), a 'ByteString' (the serialised value) and a format flag (text or
+-- binary). In our case, the format is always binary, so it really comes down
+-- to the first two: what is its type, and how do we encode it?
 type ToPrepArg :: Type -> Constraint
 class ToPrepArg a where
-  toPrepVal :: a -> PrepArg
-
-instance ToPrepArg PrepArg where
-  toPrepVal arg = arg
+  -- | Encode a value with a postgres encoder. If the value is missing (i.e.
+  -- 'Nothing'), the encoder should also be missing.
+  encoder :: a -> Maybe PE.Encoding
 
-toPrepValHelper :: PQ.Oid -> (a -> PE.Encoding) -> a -> PrepArg
-toPrepValHelper o e a = (o, Just (PE.encodingBytes $ e a, PQ.Binary))
+  -- | Get the OID for this type. Because we won't always have a value (e.g.
+  -- when we are dealing with an empty list of values), this function must be
+  -- called with a visible type parameter. For example, @'oid' @'Double'@.
+  oid :: PQ.Oid
 
 instance (ToJSON a) => ToPrepArg (AltJ a) where
-  toPrepVal (AltJ a) = toPrepValHelper PTI.JSON PE.bytea_lazy $ encode a
+  encoder (AltJ x) = pure . PE.bytea_lazy $ encode x
+  oid = PTI.JSON
+
+instance ToPrepArg a => ToPrepArg [a] where
+  encoder = pure . PE.array_foldable (fromIntegral cuint) encoder
```

Comment:
Hmmmmm..... `array_foldable` is for single dimension arrays, according to their documentation. I wonder if this works for a nested array, in which the inner type is also an array type?

Reply:
Ah, looking at it, it's fine, since we'll `encode` the inner array with `encoder` first, then fold the resulting "one-dimensional" array.

Reply:
This is my understanding - obviously there's not a lot of documentation in the module, but as far as I can tell from looking into `dimensionalArray`, this should work?